Use sidecar-shim base image with other sidecars and move them under cmd/sidecars

victortoso opened the pull request:
**What this PR does / why we need it**:

This is a follow up from the introduction of [sidecar-shim](https://quay.io/repository/kubevirt/sidecar-shim) https://github.com/kubevirt/kubevirt/pull/9231

With this PR, all hooks are implemented and all sidecars were moved to `cmd/sidecars` 

**Special notes for your reviewer**:

In this interaction we:
1. add support to `PreCloudInitIso` hook as well. Similar to `OnDefineDomain`, the `sidecar-shim` binary will load from `$PATH`  a binary named `preCloudInitIso`. The README.md was updated with more information.

2. move `cmd/example-hook-sidecar` to `cmd/sidecars/smbios` (this one is already using sidecar-shim since https://github.com/kubevirt/kubevirt/commit/d6950fb292b72f3aa9a291a264f76fadfef5fcb9

3. change `example-disk-mutation-hook-sidecar` and `example-cloudinit-hook-sidecar` to use `sidecar-shim` base image and move them under `cmd/sidecars/`  as well

4. move `cmd/network-slirp-binding` sidecar to `cmd/sidecars` without using `sidecar-shim` base image [as preferred](https://github.com/kubevirt/kubevirt/pull/10272#discussion_r1292812049). @ormergi, jfyi about moving it to `cmd/sidecars`

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
6. If no release note is required, just write "NONE".
-->
```release-note
sidecar-shim implements PreCloudInitIso hook
cloudinit and disk-mutation sidecars are now using sidecar-shim base image
```

victortoso commented:
/retest-required
/retest pull-kubevirt-e2e-arm64

victortoso commented:
/test pull-kubevirt-e2e-arm64
/test pull-kubevirt-e2e-k8s-1.27-sig-compute-migrations
/test pull-kubevirt-unit-test

aburdenthehand commented:
/cc @enp0s3

vasiliy-ul commented on cmd/sidecars/cloudinit/cloudinit.go:
```diff
@@ -0,0 +1,77 @@
+	vmi := v1.VirtualMachineInstance{}
+	err := json.Unmarshal(vmiJSON, &vmi)
+	if err != nil {
+		return "", fmt.Errorf("Failed to unmarshal given VMI spec: %s", vmiJSON)
```
nit: would it be useful to print also the actual error message? wdyt?
```suggestion
		return "", fmt.Errorf("Failed to unmarshal given VMI spec: %v %s", err, vmiJSON)
```
Same with the others `fmt.Errorf(...)` in this file and in `diskimage.go`.

victortoso replied:
Added this commit https://github.com/kubevirt/kubevirt/pull/10320/commits/8880313e4682f871ca0099517eaef72a26a77097 for `smbios` sidecar because it was missing already this fix from when I moved it to use `sidecar-shim` in d6950fb292b72f3aa9a291a264f76fadfef5fcb9. With both `cloudinit`  and `disk-mutation` sidecars, the diff is very similar  in their *use sidecar-shim* commits. Thanks for the review!

vasiliy-ul reviewed (commented):
/approve

victortoso commented on cmd/sidecars/cloudinit/BUILD.bazel:
```diff
@@ -27,22 +23,17 @@ load(
-    directory = "/",
-    entrypoint = ["/example-cloudinit-hook-sidecar"],
-    files = [":example-cloudinit-hook-sidecar"],
+    base = select({
+        "//conditions:default": "//cmd/sidecars:sidecar-shim-image",
+    }),
```
That indeed worked, at least `make bazel-generate` did not complain and `make` seems fine. I'll make this change shortly. Thanks again :)

victortoso commented:
@0xFelix and @alicefr in case you have the time to take a look :)

victortoso commented:
Rebased. No real conflicts, just underlying changes brought by https://github.com/kubevirt/kubevirt/pull/10336

victortoso commented:
/test pull-kubevirt-e2e-k8s-1.26-sig-storage

```
07:43:44:   [FAIL] [sig-storage] [Serial]ImageUpload [storage-req] Upload an image and start a VMI with PVC [test_id:4621] Should succeed [It] PVC [sig-storage, storage-req]
07:43:44:   tests/storage/imageupload.go:179 
```

victortoso commented:
/test pull-kubevirt-e2e-k8s-1.26-sig-compute
```
08:59:24:   [FAIL] [Serial][sig-compute]VirtualMachineClone Tests VM clone simple VM and cloning operations [It] clone with only some of labels/annotations [sig-compute]
08:59:24:   tests/clone_test.go:183
```

victortoso commented:
/test pull-kubevirt-e2e-k8s-1.26-sig-storage

enp0s3 commented on cmd/sidecars/smbios/smbios.go:
```diff
@@ -40,12 +40,12 @@ const (
 	domainSpec := api.DomainSpec{}
 	if err := xml.Unmarshal(domainXML, &domainSpec); err != nil {
-		return "", err
+		return "", fmt.Errorf("Failed to unmarshal given Domain spec: %s %s", err, string(domainXML))
```
I remember I had [issues](https://github.com/kubevirt/kubevirt/pull/3856) with logging the XML, therefore I had to decode it, did the log worked for you in this case?

victortoso replied:
I haven't tested this. I've added the conditional in case of errors, we have something to help debug.
The domainXML should be a valid xml string, no sure what you mean by having to decode it.

enp0s3 commented on cmd/example-disk-mutation-hook-sidecar/diskimage.go:
```diff
@@ -22,128 +22,77 @@
-	var version string
-	pflag.StringVar(&version, "version", "", "hook version to use")
+	var vmiJSON, domainXML string
+	pflag.StringVar(&vmiJSON, "vmi", "", "VMI to change in JSON format")
```
I think you've committed some temporary code here

victortoso replied:
I don't know where? Moving to sidecar-shim container image means using pflag for the binary to receive the vmi (in json format) and the domain (in xml format). I might be missing it, so let me know

enp0s3 commented on cmd/sidecars/disk-mutation/diskimage.go:
```diff
@@ -0,0 +1,98 @@
+	pflag.StringVar(&domainXML, "domain", "", "Domain spec in XML format")
+	pflag.Parse()
+
+	logger := log.New(os.Stderr, "smbios", log.Ldate)
```
`smbios` => `diskimage`

victortoso replied:
Thanks, that was a bad copy paste.

enp0s3 reviewed (commented):
@victortoso Hi! Thank you for the effort! I have some small questions below, besides that looks good to me

0xFelix commented on cmd/sidecars/cloudinit/cloudinit.go:
```diff
@@ -0,0 +1,78 @@
+	log.Print("Hook's PreCloudInitIso callback method has been called")
+
+	vmi := v1.VirtualMachineInstance{}
+	err := json.Unmarshal(vmiJSON, &vmi)
```
Inline check?

0xFelix replied:
Isn't `vmi` unused otherwise?

victortoso replied:
Indeed! This is true for the [current code](https://github.com/kubevirt/kubevirt/blob/main/cmd/example-cloudinit-hook-sidecar/cloudinit.go#L73). This example sidecar is not using it. I'll remove.

victortoso replied:
Actually, I left it for now. I'm not sure if the intention was to have it show in the example. I can do a follow up, perhaps using some `vmi` data into `cloudInitData` just to make usage of vmi... or properly remove it.

0xFelix commented on cmd/sidecars/cloudinit/cloudinit.go:
```diff
@@ -0,0 +1,78 @@
+
+	cloudInitData := cloudinit.CloudInitData{}
+	err = json.Unmarshal(cloudInitDataJSON, &cloudInitData)
+	if err != nil {
```
Same.

victortoso replied:
I usually love `inline` check but some people doesn't.
In this case, I didn't change this code, only moved it.
I don't mind changing it if you feel strong about this.

0xFelix commented on cmd/sidecars/cloudinit/cloudinit.go:
```diff
@@ -0,0 +1,78 @@
+		return "", fmt.Errorf("Failed to unmarshal given CloudInitData: %s %s", err, string(cloudInitDataJSON))
+	}
+
+	cloudInitData.UserData = "#cloud-config\n"
```
Won't this clear all of the provided `UserData`?

0xFelix replied:
To clarify, this is an example without actual use?

victortoso replied:
That's correct, just an example we have. My intentional is to move all sidecars to the same folder and use `sidecar-shim` where it is possible. I don't mind misc fixes around them. Thanks for the review!

0xFelix commented on cmd/sidecars/disk-mutation/diskimage.go:
```diff
@@ -0,0 +1,98 @@
+func OnDefineDomain(log *log.Logger, vmiJSON, domainXML []byte) (string, error) {
+	vmiSpec := vmSchema.VirtualMachineInstance{}
+	err := json.Unmarshal(vmiJSON, &vmiSpec)
+	if err != nil {
```
Inline check?

victortoso replied:
Again, this is how the code is [at the present](https://github.com/kubevirt/kubevirt/blob/main/cmd/example-disk-mutation-hook-sidecar/diskimage.go#L74), I've only applied the changes related to sidecar-shim plus moving to `cmd/sidecar` folder. I don't mind doing this changes as follow up for example.

0xFelix commented on cmd/sidecars/disk-mutation/diskimage.go:
```diff
@@ -0,0 +1,98 @@
+	}
+
+	domainSpec := domainSchema.DomainSpec{}
+	err = xml.Unmarshal(domainXML, &domainSpec)
```
Same?

victortoso replied:
[ditto](https://github.com/kubevirt/kubevirt/blob/main/cmd/example-disk-mutation-hook-sidecar/diskimage.go#L90)

0xFelix reviewed (commented):
Added some comments, otherwise I would agree with some of @enp0s3's points.

victortoso commented on cmd/sidecars/sidecar_shim.go:
```diff
@@ -54,17 +54,29 @@ type infoServer struct {
+	supportedHookPoints := map[string]string{
+		hooksInfo.OnDefineDomainHookPointName: onDefineDomainBin,
+	}
+	hookPoints := []*hooksInfo.HookPoint{}
```
Ok!

victortoso commented on cmd/sidecars/sidecar_shim.go:
```diff
@@ -54,17 +54,29 @@ type infoServer struct {
+	hookPoints := []*hooksInfo.HookPoint{}
+
+	for hookPointName, binName := range supportedHookPoints {
+		if _, err := exec.LookPath(binName); err != nil {
```
Yes, I'll be updating with conditional log, thanks!

victortoso commented on cmd/sidecars/sidecar_shim.go:
```diff
@@ -111,6 +120,35 @@ func (s v1Alpha1Server) OnDefineDomain(ctx context.Context, params *hooksV1alpha
 	}, nil
 }
 
+func runPreCloudInitIso(vmiJSON []byte, cloudInitDataJSON []byte) ([]byte, error) {
```
No. I don't think we should be `panic`. I'll change.

victortoso commented on cmd/sidecars/README.md:
```diff
@@ -19,8 +19,15 @@ communication with the main container.
 The `sidecar-shim-image` contains the `sidecar-shim` binary which should be kept as the entrypoint
```
Ok!

victortoso commented:
@0xFelix, @enp0s3 sorry for taking so long to get back to you and many thanks for the reviews!

I've rebased, that might be annoying to see the diff so I've pushed the extra patches into this [use-sidecar-shim-v2](https://github.com/victortoso/kubevirt/commits/use-sidecar-shim-v2) branch. I've squashed one commit as suggested in https://github.com/kubevirt/kubevirt/pull/10320#discussion_r1313773813, one fixup patch and all the others are in the end of the commit history.

I'm making taken suggestions as `resolved`. Let me know if I got something wrong. Cheers!

victortoso commented:
/retest-required

```
22:07:04: Summarizing 1 Failure:
22:07:04:   [FAIL] [sig-compute][Serial]CPU Hotplug A VM with cpu.maxSockets set higher than cpu.sockets [It] should successfully plug guaranteed vCPUs [sig-compute, sig-compute-migrations, requires-two-schedulable-nodes, VMLiveUpdateFeaturesGate]
22:07:04:   tests/hotplug/cpu.go:277
22:07:04: 
22:07:04: Ran 40 of 1532 Specs in 3285.425 seconds
22:07:04: FAIL! -- 39 Passed | 1 Failed | 2 Pending | 1490 Skipped
22:07:04: --- FAIL: TestTests (3285.75s)
22:07:04: FAIL
```

alicefr commented:
/lgtm
/approve
/hold

Looks good to me, let give @enp0s3 and @0xFelix some time for a last look

0xFelix reviewed (approved):
Looks good to me, thanks!

/lgtm

victortoso commented:
@enp0s3 is it fine for you? can we unhold this one? :)

victortoso commented:
Rebasing due conflicts in `hack/config-default.sh`, came from ab47865130d8d40f56831ed51b2bd3e4d8989ca0 (add's `virt-tail` in binaries)

```
<<<<<<< use-sidecar-shim
binaries_and_docker_images="
    cmd/virt-operator
    cmd/virt-controller
    cmd/virt-launcher
    cmd/virt-handler
    cmd/virt-api
    cmd/sidecars
    cmd/sidecars/smbios
    cmd/sidecars/cloudinit
"

binaries="
    ${binaries_and_docker_images}
    cmd/virtctl
    cmd/fake-qemu-process
    cmd/virt-chroot
"

docker_images="
    ${binaries_and_docker_images}
    images/disks-images-provider
    images/vm-killer
    images/nfs-server
    images/winrmcli
    tests/conformance
"

=======
binaries="cmd/virt-operator cmd/virt-controller cmd/virt-launcher cmd/virt-handler cmd/virtctl cmd/fake-qemu-process cmd/virt-api cmd/example-hook-sidecar cmd/example-cloudinit-hook-sidecar cmd/virt-chroot cmd/sidecars cmd/virt-tail"
docker_images="cmd/virt-operator cmd/virt-controller cmd/virt-launcher cmd/virt-handler cmd/virt-api images/disks-images-provider images/vm-killer images/nfs-server images/winrmcli cmd/example-hook-sidecar cmd/example-cloudinit-hook-sidecar tests/conformance cmd/sidecars"
>>>>>>> main
```

0xFelix reviewed (approved):
/lgtm

Let's get this merged

victortoso commented:
> Let's get this merged

Yep, @0xFelix we have 2 approves and 1 lgtm, should be fine to unhold it I hope? :)

0xFelix commented:
Yes, this no longer needs to be on hold!

/hold cancel

victortoso commented:
/retest-required
